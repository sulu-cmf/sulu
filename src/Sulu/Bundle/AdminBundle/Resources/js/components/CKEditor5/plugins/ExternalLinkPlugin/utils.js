// @flow
const LINK_HREF_ATTRIBUTE = 'linkHref';
const LINK_TARGET_ATTRIBUTE = 'linkTarget';

function hasExternalLinkAttribute(node: ?Object) {
    if (!node || !node.hasAttribute) {
        return false;
    }

    return node.hasAttribute(LINK_HREF_ATTRIBUTE) || node.hasAttribute(LINK_TARGET_ATTRIBUTE);
}

export {LINK_HREF_ATTRIBUTE, LINK_TARGET_ATTRIBUTE, hasExternalLinkAttribute};
