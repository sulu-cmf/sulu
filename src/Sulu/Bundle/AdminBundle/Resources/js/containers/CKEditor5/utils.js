// @flow
function addLinkConversion(editor: Object, tag: string, internalAttribute: string, tagAttribute: string) {
    editor.model.schema.extend('$text', {allowAttributes: internalAttribute});

    editor.conversion.for('upcast').attributeToAttribute({
        view: {
            name: tag,
            key: tagAttribute,
        },
        model: internalAttribute,
    });

    editor.conversion.for('downcast').attributeToElement({
        model: internalAttribute,
        view: (attributeValue, writer) => {
            return writer.createAttributeElement(tag, {[tagAttribute]: attributeValue});
        },
    });
}

function findModelItemInSelection(editor: Object) {
    const firstPosition = editor.model.document.selection.getFirstPosition();
    return firstPosition.textNode || firstPosition.nodeBefore;
}

function findViewLinkItemInSelection(editor: Object, linkTag: string) {
    const selection = editor.editing.view.document.selection;
    const firstPosition = selection.getFirstPosition();

    return firstPosition.getAncestors().find(
        (ancestor) => ancestor.is('attributeElement') && ancestor.name === linkTag
    );
}

export {addLinkConversion, findModelItemInSelection, findViewLinkItemInSelection};
