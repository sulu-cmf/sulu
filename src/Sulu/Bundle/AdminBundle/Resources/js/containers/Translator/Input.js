// @flow
import React from 'react';
import {TextEditor} from '../../containers';
import translatorStyles from './translator.scss';

type Props = {|
    onChange?: (text: string) => void,
    text: string,
    type: 'text_line' | 'text_area' | 'text_editor',
|};

/**
 * @internal
 */
export default class Input extends React.Component<Props> {
    handleChange = (value: ?string) => {
        const {
            onChange,
        } = this.props;

        if (onChange) {
            onChange(value || '');
        }
    };

    handleTextAreChange = (event: SyntheticEvent<HTMLInputElement>) => {
        this.handleChange(event.currentTarget.value);
    };

    render() {
        const {
            type,
        } = this.props;

        if (type === 'text_editor') {
            return this.renderEditor();
        }

        return this.renderTextarea();
    }

    renderEditor() {
        const {
            text,
            onChange,
        } = this.props;

        return (
            <div className={translatorStyles.input + ' ' + translatorStyles.texteditor}>
                <TextEditor
                    adapter="ckeditor5"
                    disabled={onChange === undefined}
                    locale={undefined}
                    onChange={this.handleChange}
                    value={text}
                />
            </div>
        );
    }

    renderTextarea() {
        const {
            text,
        } = this.props;

        return (
            <textarea
                className={translatorStyles.input + ' ' + translatorStyles.textarea}
                onChange={this.handleTextAreChange}
                value={text}
            />
        );
    }
}
