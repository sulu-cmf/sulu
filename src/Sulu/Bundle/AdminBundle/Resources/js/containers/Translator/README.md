The `Translator` container can be used to display a deepl like overlay.

> **_NOTE:_** This container is experimental can be changed in the future and is not covered by our BC promise.
