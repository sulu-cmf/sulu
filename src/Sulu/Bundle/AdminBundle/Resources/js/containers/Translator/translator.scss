.translator {
    display: flex;
    margin: 40px 50px;
    background-color: #fff;
    border: 1px solid #ccc;
}

.column {
    flex: 1;
    display: flex;
    flex-direction: column;
    max-width: 50%;

    &:first-child {
        border-right: 1px solid #ccc;
    }
}

.select {
    max-width: 50%;
    height: 37px;
    padding-top: 10px;
}

.input {
    margin-top: 10px;
    width: 100%;
    height: 200px;
    resize: vertical;
    border-top: 1px solid #ccc;
    background-color: #fff;
}

.textarea {
    resize: none;
    outline: none;
    border-left: 0;
    border-right: 0;
    border-bottom: 0;
    padding: 10px;
}

.texteditor {
    :global(.ck-editor__top) {
        display: none;
    }

    :global(.ck-editor__editable),
    :global(.ck.ck-editor .ck.ck-content.ck-focused) {
        border: 0 !important;
        height: 200px;
    }

    :global(.ck.ck-editor .ck.ck-content) {
        height: 199px;
    }

    :global(.ck.disabled) {
        pointer-events: all !important;
    }
}
