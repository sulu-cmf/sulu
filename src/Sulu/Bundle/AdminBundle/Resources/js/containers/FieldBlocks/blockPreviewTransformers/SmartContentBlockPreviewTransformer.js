// @flow
import React from 'react';
import type {Node} from 'react';
import type {BlockPreviewTransformer} from '../types';
import {translate} from '../../../utils/Translator';

export default class SmartContentBlockPreviewTransformer implements BlockPreviewTransformer {
    transform(value: *): Node {
        return (
            <p>
                <em>
                    {translate(
                        'sulu_admin.smart_content_block_preview',
                        {limit: value.limitResult ? value.limitResult : 'undefined'}
                    )}
                </em>
            </p>
        );
    }
}
