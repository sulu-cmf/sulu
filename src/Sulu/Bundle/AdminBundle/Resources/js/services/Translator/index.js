// @flow
import {setTranslations, clearTranslations, translate} from './Translator';

export {setTranslations, clearTranslations, translate};
