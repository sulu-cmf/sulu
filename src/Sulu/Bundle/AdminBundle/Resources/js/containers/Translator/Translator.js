// @flow

import React from 'react';
import {observer} from 'mobx-react';
import debounce from 'debounce';
import {action, observable, toJS} from 'mobx';
import {Overlay, SingleSelect} from '../../components';
import {Requester} from '../../services';
import translatorStyles from './translator.scss';
import Input from './Input';

type Props = {|
    action?: React$ComponentType<{|
        context: Object,
        source: string,
    |}>,
    locale: string,
    messages: {|
        detected: string,
        errorTranslatingText: string,
        insert: string,
        title: string,
    |},
    onConfirm: (text: string) => void,
    onDialogClose: () => void,
    sourceLanguages: Array<{|
        label: string,
        locale: string,
    |}>,
    targetLanguages: Array<{|
        label: string,
        locale: string,
    |}>,
    type: 'text_line' | 'text_area' | 'text_editor',
    url: string,
    value?: string,
|};

/**
 * @internal
 */
@observer
export default class Translator extends React.Component<Props> {
    @observable snackbarMessage: ?{ message: string, type: 'error' } = undefined;
    @observable loading = false;
    @observable sourceText = '';
    @observable targetText = '';
    @observable sourceLanguage = undefined;
    @observable sourceSelectedOnce = false;
    @observable targetLanguage = undefined;
    @observable lastResponse = undefined;

    @action handleClose = () => {
        const {onDialogClose} = this.props;

        onDialogClose();
    };

    @action handleConfirm = () => {
        const {
            onConfirm,
        } = this.props;

        // We have to stop the propagation of the event to prevent the focus lose of the input / editor field
        // $FlowFixMe
        event.stopPropagation();

        onConfirm(this.targetText);
    };

    @action handleSnackbarCloseClick = () => {
        this.snackbarMessage = undefined;
    };

    @action componentDidMount() {
        this.targetLanguage = this.props.locale;
        this.sourceText = this.props.value;

        this.translateText(this.sourceText);
    }

    @action handleSourceTextChanged = (text: string) => {
        this.sourceText = text;

        this.translateText(text);
    };

    translateText = debounce(action(() => {
        const {
            url,
            messages: {
                errorTranslatingText: errorTranslatingTextMessage,
            },
        } = this.props;

        this.loading = true;
        this.lastResponse = undefined;

        return Requester.post(
            url,
            {
                text: this.sourceText,
                sourceLanguage: this.sourceLanguage,
                targetLanguage: this.targetLanguage,
            }
        ).then(action((data: {
            response: {
                sourceLanguage: string,
                targetLanguage: string,
                text: string,
            },
        }) => {
            this.loading = false;
            this.targetText = data.response.text;
            this.lastResponse = data;

            this.targetLanguage = data.response.targetLanguage.toLowerCase();
            this.sourceLanguage = data.response.sourceLanguage.toLowerCase();

            return data;
        })).catch(action((error) => {
            this.loading = false;
            this.lastResponse = {error};

            this.snackbarMessage = {
                message: errorTranslatingTextMessage,
                type: 'error',
            };
        }));
    }), 500);

    @action handleSourceLanguageChanged = (locale: string) => {
        this.sourceLanguage = locale;
        this.sourceSelectedOnce = true;

        this.translateText(this.sourceText);
    };

    @action handleTargetLanguageChanged = (locale: string) => {
        this.targetLanguage = locale;

        this.translateText(this.sourceText);
    };

    render() {
        const {
            type,
            sourceLanguages,
            targetLanguages,
            action: Action,
            messages: {
                title: titleMessage,
                insert: insertMessage,
                detected: detectedMessage,
            },
        } = this.props;

        const actionNode = Action ? (
            <Action
                context={toJS(this.lastResponse)}
                source="translator"
            />
        ) : <React.Fragment />;

        return (
            <Overlay
                confirmDisabled={this.targetText === ''}
                confirmLoading={this.loading}
                confirmText={insertMessage}
                onClose={this.handleClose}
                onConfirm={this.handleConfirm}
                onSnackbarCloseClick={this.handleSnackbarCloseClick}
                open={true}
                size="small"
                snackbarMessage={this.snackbarMessage?.message}
                snackbarType={this.snackbarMessage?.type}
                title={titleMessage}
            >
                {actionNode}

                <div className={translatorStyles.translator}>
                    <div className={translatorStyles.column}>
                        <div className={translatorStyles.select}>
                            <SingleSelect
                                onChange={this.handleSourceLanguageChanged}
                                skin="flat"
                                value={this.sourceLanguage}
                            >
                                {sourceLanguages.map((option) => {
                                    const isDetected = option.locale.toLowerCase() === this.sourceLanguage
                                        && !this.sourceSelectedOnce;

                                    return (
                                        <SingleSelect.Option key={option.locale} value={option.locale.toLowerCase()}>
                                            {option.label}
                                            {isDetected && ' (' + detectedMessage + ')'}
                                        </SingleSelect.Option>
                                    );
                                })}
                            </SingleSelect>
                        </div>
                        <Input
                            onChange={this.handleSourceTextChanged}
                            text={this.sourceText || ''}
                            type={type}
                        />
                    </div>
                    <div className={translatorStyles.column}>
                        <div className={translatorStyles.select}>
                            <SingleSelect
                                onChange={this.handleTargetLanguageChanged}
                                skin="flat"
                                value={this.targetLanguage}
                            >
                                {targetLanguages.map((option) => (
                                    <SingleSelect.Option key={option.locale} value={option.locale.toLowerCase()}>
                                        {option.label}
                                    </SingleSelect.Option>
                                ))}
                            </SingleSelect>
                        </div>
                        <Input
                            text={this.targetText}
                            type={type}
                        />
                    </div>
                </div>
            </Overlay>
        );
    }
}
