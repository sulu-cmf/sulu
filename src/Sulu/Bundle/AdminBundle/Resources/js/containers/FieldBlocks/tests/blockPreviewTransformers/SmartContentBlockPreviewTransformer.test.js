// @flow
import SmartContentBlockPreviewTransformer from '../../blockPreviewTransformers/SmartContentBlockPreviewTransformer';
import {translate} from '../../../../utils/Translator';

jest.mock('../../../../utils/Translator', () => ({
    translate: jest.fn((key) => key),
}));

test('Return JSX for configuration with a limit of 5', () => {
    const smartContentBlockPreviewTransformer = new SmartContentBlockPreviewTransformer();
    expect(smartContentBlockPreviewTransformer.transform({limitResult: 5})).toMatchSnapshot();
    expect(translate).toBeCalledWith('sulu_admin.smart_content_block_preview', {limit: 5});
});

test('Return null for everything expect a string', () => {
    const smartContentBlockPreviewTransformer = new SmartContentBlockPreviewTransformer();
    expect(smartContentBlockPreviewTransformer.transform({})).toMatchSnapshot();
    expect(translate).toBeCalledWith('sulu_admin.smart_content_block_preview', {limit: 'undefined'});
});
