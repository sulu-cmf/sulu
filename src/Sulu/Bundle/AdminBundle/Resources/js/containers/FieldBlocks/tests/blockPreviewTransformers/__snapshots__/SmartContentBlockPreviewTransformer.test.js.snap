// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Return JSX for configuration with a limit of 5 1`] = `
<p>
  <em>
    sulu_admin.smart_content_block_preview
  </em>
</p>
`;

exports[`Return null for everything expect a string 1`] = `
<p>
  <em>
    sulu_admin.smart_content_block_preview
  </em>
</p>
`;
